federation/server.c

```diff
@@ -0,0 +1,765 @@
+        arrfree(server->bundle_tds[idx].value);
+
+        shdel(server->bundle_sources, path);
+        shdel(server->bundle_tds, path);
+    }
+    mtx_unlock(&server->mutex);
+
+    return NO_ERROR;
+}
+
+// load keys to use with 'https_web'
+// register a HTTPS_WEB endpoint, for starting with
+// spiffebundle_EndpointServer_ServeEndpoint
+spiffebundle_EndpointInfo *spiffebundle_EndpointServer_AddHttpsWebEndpoint(
+    spiffebundle_EndpointServer *server, const char *base_url, X509 **certs,
+    EVP_PKEY *priv_key, err_t *error)
+{
+    if(!server) {
+        *error = ERR_NULL;
+        return NULL;
+    }
+    if(!base_url) {
+        *error = ERR_BAD_ARGUMENT;
+        return NULL;
+    }
+    if(!certs || !(certs[0])) {
+        *error = ERR_CERTIFICATE_VALIDATION;
+        return NULL;
+    }
+    priv_key = x509svid_validatePrivateKey(priv_key, certs[0], error);
+    if(!priv_key) {
+        *error = ERR_PRIVKEY_VALIDATION;
+        return NULL;
+    }
+
+    mtx_lock(&server->mutex);
+    int idx = shgeti(server->endpoints, base_url);
+    if(idx >= 0) {
+        EVP_PKEY_free(priv_key);
+        mtx_unlock(&server->mutex);
+        *error = ERR_EXISTS;
+        return NULL;
+    }
+
+    spiffebundle_EndpointInfo *e_info = spiffebundle_EndpointInfo_New();
+    mtx_lock(&e_info->mutex);
+    shput(server->endpoints, base_url, e_info);
+    e_info->server = server;
+    mtx_unlock(&server->mutex);
+
+    // create svid with blank spiffeid for holding the key and certificates
+    x509svid_SVID *svid = calloc(1, sizeof *svid);
+    for(size_t i = 0, size = arrlenu(certs); i < size; ++i) {
+        X509_up_ref(certs[i]);
+        arrput(svid->certs, certs[i]);
+    }
+    svid->private_key = priv_key;
+
+    x509svid_Source *source = x509svid_SourceFromSVID(svid);
+    e_info->listen_mode = spiffetls_TLSServerWithRawConfig(source);
+    e_info->url = string_new(base_url);
+    e_info->threads = NULL;
+    mtx_unlock(&e_info->mutex);
+
+    *error = NO_ERROR;
+    return e_info;
+}
+
+err_t spiffebundle_EndpointServer_SetHttpsWebEndpointAuth(
+    spiffebundle_EndpointServer *server, const char *base_url, X509 **certs,
+    EVP_PKEY *priv_key)
+{
+    if(!server) {
+        return ERR_NULL;
+    }
+    if(!base_url) {
+        return ERR_BAD_ARGUMENT;
+    }
+    if(!certs || !(certs[0])) {
+        return ERR_CERTIFICATE_VALIDATION;
+    }
+    err_t error = NO_ERROR;
+    priv_key = x509svid_validatePrivateKey(priv_key, certs[0], &error);
+    if(!priv_key) {
+        return ERR_PRIVKEY_VALIDATION;
+    }
+    mtx_lock(&server->mutex);
+    int idx = shgeti(server->endpoints, base_url);
+    if(idx < 0) {
+        EVP_PKEY_free(priv_key);
+        mtx_unlock(&server->mutex);
+        return ERR_EXISTS;
+    }
+    spiffebundle_EndpointInfo *e_info = server->endpoints[idx].value;
+    mtx_lock(&e_info->mutex);
+
+    spiffetls_ListenMode_Free(e_info->listen_mode);
+
+    x509svid_SVID *svid = calloc(1, sizeof *svid);
```

Comment:
Use x509svid_newSVID function

Reply:
x509svid_newSVID() checks for a valid x509 SVID, with an spiffeid in it.
This is a "fake"/"manufactured" x509 SVID for holding the TLS certificate and private key for the server.
This way we can use the certificate and key with spiffetls.